dace/runtime/include/dace/math.h

```diff
@@ -512,40 +512,84 @@ namespace dace
             return (thrust::complex<T>)thrust::pow(a, b);
         }
 #endif
-        template<typename T, typename U>
+
+
+/* If `DACE_XILINX` is defined then `pow` only returns floats, even if their arguments
+ *  are integer. If it is not defined, then if all arguments of `pow` are integers
+ *  the return value will also be an integer, currently it is always the largest supported by the platform.
+ *  This is the original behaviour prior to [PR#1748](https://github.com/spcl/dace/pull/1748). */
+#if defined(DACE_XILINX)
+
+        template<
+            typename T,
+            typename U
+        >
         DACE_CONSTEXPR DACE_HDFI auto pow(const T& a, const U& b)
         {
             return std::pow(a, b);
         }
 
-#ifndef DACE_XILINX
-        static DACE_CONSTEXPR DACE_HDFI int pow(const int& a, const int& b)
+#else
+
+        template<
+            typename T,
+            typename U,
+            typename = std::enable_if_t<!(std::is_integral<T>::value && std::is_integral<U>::value)>
+        >
+        DACE_CONSTEXPR DACE_HDFI auto pow(const T& a, const U& b)
         {
-            if (b < 0) return 0;
-            int result = 1;
-            for (int i = 0; i < b; ++i)
-                result *= a;
-            return result;
+            return std::pow(a, b);
         }
 
-        static DACE_CONSTEXPR DACE_HDFI unsigned int pow(const unsigned int& a,
-                                       const unsigned int& b)
+
+        //TODO: Should this always be the largest integer?
+        template<typename T>
+        using IntPowReturnType_t = std::conditional_t<std::is_unsigned<T>::value, uintmax, intmax>;
+
+
+       /* TODO: The return value is always an integer, this is different from the behaviour of `std::pow` that
+        *  always return a float. We should probably patch the code generator to generate `ipow` calls if all
+        *  arguments are integers. */
+        template<
+            typename T1,
+            typename T2,
+            typename = std::enable_if_t<std::is_integral<T1>::value && std::is_integral<T2>::value>
```

Comment:
This is correct.
But we kind of need this behaviour, to enable the FFT test this commit enables.
There the problem is that a size of an array is given as `a = np.ones(2**3)`.
Which the code generator translates to
```cpp
double* a;
a = new double[pow(2, 3)];
```
If `pow` would return a double, then it is a compilation error.

There are technically three things we can do:
- Leave it as it is.
- Adding the specialization that if the exponent is an unsigned int then we return a double and hope that expressions such as above are never called where the exponent is unsigned, i.e.
```cpp
unsigned int b = 4;
//...
double* a;
a = new double[pow(2, b)];
```
- Patch the code generator that sizes in `new[]` expressions are always casted to `std::size_t`.
However, this would be another PR.

What is your opinion on this @tbennun ?

Reply:
I agree with @alexnick83. The reason we didn't have these pow implementations in the first place was because of negative exponents. I think we should use `ipow` when we know something or the equivalent version only in the case of an unsigned exponent and integer base.

Reply:
I looked it up in the code generator, `ipow` is only used if the exponent is known at code generation time.
This is not the case in the FFT test, there the exponent is not known and  thus `pow` is selected.

The reason why it lead to the compilation error in the FFT test was because the arguments of `pow` where `int64` and since there is only an overload for `int` and `unsigned` int, the template was selected.
However, even before the expression `5^-3` would have resulted in the value 0, given that both were integer.

For the reasons outlined above, I do not think that this PR changes anything big.
But how should I proceed @alexnick83 @tbennun 

PS: I also tried to implement `ipow` for negative exponents and that failed with some compilation bug related to `half`, which I did not understood.